app/src/main/java/org/sagebionetworks/research/mpower/ActivityRecorderService.kt

```diff
@@ -0,0 +1,280 @@
+//                    delay(MAX_RECORDING_TIME_MS)
+//                    stopRecording()
+//                }
+                // NOTE: Currently RecorderService within RecorderManager is not bound, need to fix
+                // don't care about navDirection
+                recorderManager.onStepTransition(null, startStep, 0)
+            } else {
+                Log.d(TAG, "-- NOT RECORDING - ONLY RECORD ONCE EVERY $MIN_FREQUENCY_MS")
+                stopSelf()
+            }
+        }
+    }
+
+    private fun startForeground() {
+        createNotificationChannel()
+        val notification: Notification = NotificationCompat.Builder(this, getString(R.string.foreground_channel_id))
+            .setContentTitle(getText(R.string.recording_notification_title))
+            .setContentText(getText(R.string.recording_notification_message))
+            .setSmallIcon(R.drawable.ic_launcher_foreground)
+            .build()
+        startForeground(FOREGROUND_NOTIFICATION_ID, notification)
+    }
+
+    private fun createNotificationChannel() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            // Create the NotificationChannel
+            val title = getString(R.string.foreground_channel_title)
+            val desc = getString(R.string.foreground_channel_desc)
+            val importance = NotificationManager.IMPORTANCE_LOW
+            val mChannel = NotificationChannel(getString(R.string.foreground_channel_id), title, importance)
+            mChannel.description = desc
+            // Register the channel with the system; you can't change the importance
+            // or other notification behaviors after this
+            val notificationManager = getSystemService(NOTIFICATION_SERVICE) as NotificationManager
+            notificationManager.createNotificationChannel(mChannel)
+        }
+    }
+
+    //WALK END EVENT - Ignore if recording is already done
+    // Should not need to call startForeground as service is either already running or it wasn't and
+    // this is just a NO-OP and a call to stopService.
+    // Stop recording
+    // Save to bridge - any recording is worth saving
+    // Stop service
+    private fun stopRecording() {
+        if (isRecording) {
+            Log.d(
+                TAG,
+                "-- stopRecording ${SimpleDateFormat(TIME_PATTERN, Locale.US).format(Date())}"
+            )
+            isRecording = false
+//            CoroutineScope(Dispatchers.Default).launch {
+//                recorderService.getActiveRecorders(taskUUID!!).forEach {
+//                    it.value.stop()
+//                }
+//            }
+            recorderManager.onStepTransition(stopStep, null, 0)
+
+            val sharedPrefs = getSharedPreferences(TRANSITION_PREFS, Context.MODE_PRIVATE)
+            sharedPrefs.edit().putLong(LAST_RECORDED_AT, Date().time).apply()
+        }
+        stopSelf()
+    }
+
+    override fun onBind(intent: Intent?): IBinder? {
+        return null
```

Comment:
What do we get by returning null here (and could this by chance be our null problem?)

Reply:
services can be optionally bounded by a component. returning null just means this particular service is not bound by any android component.  very normal to return null here.